### DISCUSSION TITLE
How to hide the startup window?

### DISCUSSION THREAD

**Discussion by jet10000** [flet-dev/flet#353] [General]
https://user-images.githubusercontent.com/2258120/191020203-3061efc7-c5a1-410a-880b-3e86a6c13553.mp4

<img width="513" alt="image" src="https://user-images.githubusercontent.com/2258120/191020545-feb6b99e-1bd7-4abd-9f26-257a3801ccdc.png">

Always have a window flashing in the middle of the screen at startup

**Comment by FeodorFitsner** [comment]
You mean this: https://flet.dev/docs/controls/page#window_visible?

**Reply by jet10000** [reply]
I tried it and it doesn't solve the problem.

**Comment by FeodorFitsner** [comment]
What OS are you running on?

**Reply by jet10000** [reply]
<img width="241" alt="image" src="https://user-images.githubusercontent.com/2258120/191067896-bb8019ea-d17a-460c-9a98-b08c244ff2d6.png">

**Reply by jet10000** [reply]
I tested the window_manager plugin in flutter, which can be hidden, avoiding showing it in the middle of the screen first, and then jumping to the right side of the screen to show it.

**Reply by jet10000** [reply]
<img width="563" alt="image" src="https://user-images.githubusercontent.com/2258120/191068596-e61e1c36-d2dd-49a6-aa6d-37f1a8e1c70e.png">

**Reply by FeodorFitsner** [reply]
You mean your Flutter sample worked, but Flet equivalent didn't?

**Reply by jet10000** [reply]
yes

**Comment by FeodorFitsner** [comment]
Can you post complete both Flet and Flutter examples as text?

**Comment by jet10000** [comment]
The following is the main code of the two examples

# flutter

```dart
Future setWindowPosition(double? top, double? left) async {
  if (Platform.isMacOS) {
    // debugPrint("setWindowPosition()");
    // var currentPos = await windowManager.getPosition();
    // if (kDebugMode) {
    //   print(currentPos);
    // }
    await windowManager.setPosition(
        // Offset(left ?? currentPos.dx, top ?? currentPos.dy),
        const Offset(1050.0, 100.0),
        animate: defaultTargetPlatform != TargetPlatform.macOS);
  }
}

void main() async {
  if (Platform.isMacOS) {
    // for macos app
    WidgetsFlutterBinding.ensureInitialized();
    // Must add this line.
    await windowManager.ensureInitialized();

    WindowOptions windowOptions = const WindowOptions(
      size: Size(350, 680),
      // center: true,
      // backgroundColor: Colors.transparent,
      skipTaskbar: false,
      // titleBarStyle: TitleBarStyle.hidden,
      title: '',
      alwaysOnTop: true,
    );
    windowManager.waitUntilReadyToShow(windowOptions, () async {
      await setWindowPosition(1000.0, 100.0);
      await windowManager.show();
      await windowManager.focus();
    });
  }
  runApp(const MyApp());
}
```

# flet

```python
def main(page: Page):
    # page.title = "X-Shot"
    # page.vertical_alignment = "center"
    page.window_width = 300
    page.window_top = 130
    page.window_left = 1100
    page.window_always_on_top = True
    page.scroll = True
    page.theme = theme.Theme(color_scheme_seed="green")
    # page.theme.use_material3 = True

    def copyit(e):
        page.update()

    def saveit(e):
    		page.update()

    def test123(e):
        print('afa')

    page.add(
        Row([
            ElevatedButton(text="Copy to Board", on_click=copyit),
            ElevatedButton(text="Save", on_click=saveit),
        ],
            alignment="center",
        )
    )


# flet.app(target=main, view=flet.WEB_BROWSER)
flet.app(target=main)
```
```

**Reply by FeodorFitsner** [reply]
To hide window on startup Flet app should be started as:

```
flet.app(target=main, view=flet.FLET_APP_HIDDEN)
```

note `view=flet.FLET_APP_HIDDEN`.

Then you should make the window visible:

```
    page.window_visible = True
    page.update() 
```

Try this full example:

```python
def main(page: Page):
    # page.title = "X-Shot"
    # page.vertical_alignment = "center"
    page.window_width = 300
    page.window_top = 130
    page.window_left = 1100
    page.window_always_on_top = True
    page.scroll = True
    page.theme = theme.Theme(color_scheme_seed="green")
    # page.theme.use_material3 = True
    page.window_visible = True
    page.update() 

    def copyit(e):
        page.update()

    def saveit(e):
    		page.update()

    def test123(e):
        print('afa')

    page.add(
        Row([
            ElevatedButton(text="Copy to Board", on_click=copyit),
            ElevatedButton(text="Save", on_click=saveit),
        ],
            alignment="center",
        )
    )


flet.app(target=main, view=flet.FLET_APP_HIDDEN)
```

**Reply by jet10000** [reply]
Nice

**Comment by FeodorFitsner** [comment]
Did it work?

**Reply by jet10000** [reply]
It's worked. 🍻

**Reply by jet10000** [reply]
if run `pipenv run flet main.py `, the  window flashing display again. if run `pipenv run python main.py`, it's worked.

**Comment by FeodorFitsner** [comment]
Use `-n` (`--hidden`) parameter:

    pipenv run flet main.py -n

**Reply by AthenaRoland** [reply]
How to hide the app window instead of exiting with the close button？

**Reply by feb-cloud** [reply]
@FeodorFitsner 

My test environment:
`Python 3.8.13`
`Windows-10-10.0.19045-SP0`

Tested Flet versions: Every release from `0.23.2 - 0.15.0`.

Test code:

```python
from time import sleep

import flet as ft

def main(page: ft.Page):
page.add(ft.Text("Hello!"))

sleep(3)

page.window_visible = True
page.update()

ft.app(main, view=ft.AppView.FLET_APP_HIDDEN)
````

Conclusion: `view=ft.AppView.FLET_APP_HIDDEN` is no longer effective since `0.16.0`, but it is effective in `0.15.0`.

I tested another version to be released, `0.24.0.dev3317`, and the problem still exists.

**Comment by hschneider** [comment]
This workaround does not work on Flet >0.80 anymore.
Any hints?

**Comment by hschneider** [comment]
Got it!
```python
    page.window.visible = True
    page.window.update()
...
ft.run(main, view=ft.AppView.FLET_APP_HIDDEN)
```
`page.window.update()` did the trick. No `sleep(3)` required.

**Comment by hschneider** [comment]
OK, this works like a charm when using `flet run -r`, but not with the compiled app. 
When starting the macOS App Bundle, the small white window shows up again.

**Reply by ndonkoHenri** [reply]
See this: https://flet.dev/docs/publish/#hidden-app-window-on-startup

**Reply by hschneider** [reply]
Setting
```
[tool.flet.app]    # or [tool.flet.<PLATFORM>.app]
hide_window_on_start = true
```
and 
```
    page.window.visible = True
    page.update()

ft.run(main, view=ft.AppView.FLET_APP_HIDDEN)
```
results in a permanently hidden window under macOS.
Nothing shows up at all.
